### PR TITLE
[Improvement][API] Support to check if the worker group has been used by any tasks or schedulers when users delete or rename it. 

### LINKED ISSUE #14864: [Improvement][UI] Need to check if the worker group has been used by any tasks or schedulers when users delete or rename it.
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

At present, the DS has used the name of worker group as the unique identifier. And all of tasks and schedulers use its name in their settings. Once it's deleted or renamed, its dependent tasks won't be able to work well. So when users delete or rename  a worker group, it's supposed to check if it has been used by any tasks or schedulers.

This issue is related to #11805.

image

image

image

image

image

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by calvinjiang**
## Purpose of the pull request

This PR will close #14864 .

## Verify this pull request
This change added tests and can be verified as follows:
- *Manually verified the change by testing locally.*

The test results are as follows:

image

image

### CODE REVIEW

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java [hunk 1]
```diff
@@ -177,6 +195,52 @@ private boolean checkWorkerGroupNameExists(WorkerGroup workerGroup) {
...
+     * @param workerGroup worker group
+     * @return boolean
+     */
+    private boolean checkWorkerGroupDependencies(WorkerGroup workerGroup, Map<String, Object> result) {
+        // check if the worker group has any dependent tasks
+        List<TaskDefinition> taskDefinitions = taskDefinitionMapper.selectList(
+                new QueryWrapper<TaskDefinition>().lambda().eq(TaskDefinition::getWorkerGroup, workerGroup.getName()));
+
+        if (CollectionUtils.isNotEmpty(taskDefinitions)) {
+            List<String> taskNames = taskDefinitions.stream().limit(3).map(taskDefinition -> taskDefinition.getName())
```
Comment (zhuangchong) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java:210]: Only fetch 3 records?

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java [hunk 2]
```diff
@@ -177,6 +195,52 @@ private boolean checkWorkerGroupNameExists(WorkerGroup workerGroup) {
...
+    /**
+     * check if the worker group has any dependent tasks,schedulers or environments.
+     *
+     * @param workerGroup worker group
+     * @return boolean
+     */
+    private boolean checkWorkerGroupDependencies(WorkerGroup workerGroup, Map<String, Object> result) {
+        // check if the worker group has any dependent tasks
+        List<TaskDefinition> taskDefinitions = taskDefinitionMapper.selectList(
+                new QueryWrapper<TaskDefinition>().lambda().eq(TaskDefinition::getWorkerGroup, workerGroup.getName()));
```
Comment (zhuangchong) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java:207]: Should we add task online status conditions?

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java [hunk 3]
```diff
@@ -177,6 +195,52 @@ private boolean checkWorkerGroupNameExists(WorkerGroup workerGroup) {
...
+                    .collect(Collectors.toList());
+
+            putMsg(result, Status.WORKER_GROUP_DEPENDENT_TASK_EXISTS, taskDefinitions.size(),
+                    JSONUtils.toJsonString(taskNames));
+            return true;
+        }
+
+        // check if the worker group has any dependent schedulers
+        List<Schedule> schedules = scheduleMapper
+                .selectList(new QueryWrapper<Schedule>().lambda().eq(Schedule::getWorkerGroup, workerGroup.getName()));
```
Comment (zhuangchong) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java:220]: Should we add schedule online status conditions?

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java [hunk 4]
```diff
@@ -177,6 +195,52 @@ private boolean checkWorkerGroupNameExists(WorkerGroup workerGroup) {
...
+
+        if (CollectionUtils.isNotEmpty(schedules)) {
+            List<String> processNames = schedules.stream().limit(3)
+                    .map(schedule -> processDefinitionMapper.queryByCode(schedule.getProcessDefinitionCode()).getName())
+                    .collect(Collectors.toList());
+
+            putMsg(result, Status.WORKER_GROUP_DEPENDENT_SCHEDULER_EXISTS, schedules.size(), processNames);
```
Comment (zhuangchong) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java]: processNames? The above taskNames use JSON serialization, please choose a method and keep it consistent.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java (hunk 1 above)
Comment (calvinjiang) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java:210]: Here are just a part of examples to help users understand why they can not remove or rename the work group. If there are a lot of tasks, timings or environments that are dependent on the work group, it's not so good for the tip to display all of those.

**Review by zhuangchong — APPROVED** [review]
+1

**Comment by xjlgod** [comment]
I see SonarCloud check and OWASP did not pass. Should I add more unit test? I think  OWASP is not about my pr.

**Pull request merged**